README
X86-Disasm version 0.02
=======================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires libdisasm - a C disassembler available from 
http://bastard.sourceforge.net/libdisasm.html

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Bob Wilkinson

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension X86::Disasm.

0.01  Thu Jun 11 20:57:03 2009
	- original version; created by h2xs 1.23 with options
		-A -a -n X86::Disasm /usr/include/libdis.h

0.02  Fri Nov 19 20:10:25 2010
        - add checks to Makefile.PL for C library and refuse to build unless it exists


Makefile.PL
use 5.008000;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'X86::Disasm',
    VERSION_FROM      => 'lib/X86/Disasm.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/X86/Disasm.pm', # retrieve abstract from module
       AUTHOR         => 'Bob Wilkinson <bob@>') : ()),
    LIBS              => ['-ldisasm'], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/X86/Disasm.pm
package X86::Disasm;

use 5.008000;
use strict;
use warnings;

use X86::Disasm::Insn ':all';

require Exporter;

our @ISA = qw(Exporter);

# automagically read these from /usr/include/libdis.h
# with a parser - all hashes but first

our $x86_report_codes = {
        'report_disasm_bounds',   #  RVA OUT OF BOUNDS : The disassembler could
                                  #  not disassemble the supplied RVA as it is
                                  #  out of the range of the buffer. The
                                  #  application should store the address and
                                  #  attempt to determine what section of the
                                  #  binary it is in, then disassemble the
                                  #  address from the bytes in that section.
                                  #       data: uint32_t rva
        'report_insn_bounds',     #  INSTRUCTION OUT OF BOUNDS: The disassembler
                                  #  could not disassemble the instruction as
                                  #  the instruction would require bytes beyond
                                  #  the end of the current buffer. This usually
                                  #  indicated garbage bytes at the end of a
                                  #  buffer, or an incorrectly-sized buffer.
                                  #       data: uint32_t rva
        'report_invalid_insn',    #* INVALID INSTRUCTION: The disassembler could
                                  #  not disassemble the instruction as it has an
                                  #  invalid combination of opcodes and operands.
                                  #  This will stop automated disassembly; the
                                  #  application can restart the disassembly
                                  #  after the invalid instruction.
                                  #       data: uint32_t rva
        'report_unknown'
};

our $x86_options = {	# these can be ORed together 
	0 => "opt_none",
	1 => "opt_ignore_nulls",	 # ignore sequences of > 4 NULL bytes
	2 => "opt_16_bit",	 # 16-bit/DOS disassembly
	4 => "opt_att_mnemonics",	 # use AT&T syntax names for alternate opcode mnemonics
};

our $x86_op_foreach_type = {
	0 => "op_any",	 # ALL operands (explicit, implicit, rwx)
	1 => "op_dest",	 # operands with Write access
	2 => "op_src",	 # operands with Read access
	3 => "op_ro",	 # operands with Read but not Write access
	4 => "op_wo",	 # operands with Write but not Read access
	5 => "op_xo",	 # operands with Execute access
	6 => "op_rw",	 # operands with Read AND Write access
	0x10 => "op_implicit",	 # operands that are implied by the opcode
	0x20 => "op_explicit",	 # operands that are not side-effects
};

our $x86_asm_format_enum = {
	"unknown_syntax" => 0,	 # never use!
        "native_syntax"  => 1,   # header: 35 bytes 
        "intel_syntax"   => 2,   # header: 23 bytes
        "att_syntax"     => 3,   # header: 23 bytes
        "xml_syntax"     => 4,   # header: 679 bytes
        "raw_syntax"     => 5    # header: 172 bytes
};

our $x86_asm_format = {
	0 => "unknown_syntax",	 # never use!
        1 => "native_syntax",    # header: 35 bytes
        2 => "intel_syntax",     # header: 23 bytes
        3 => "att_syntax",       # header: 23 bytes
        4 => "xml_syntax",       # header: 679 bytes
        5 => "raw_syntax"        # header: 172 bytes
};

sub new {
  my $proto = shift;
  my $class = ref($proto) || $proto;
  my $self = {};
  bless $self, $class;

#The first argument to x86_init() represents disassembler options; these are 
#defined as our $x86_options above

  my $options = @_ ? shift : 0;
  my $reporter = @_ ? shift : 0;
#use Devel::Peek; Dump($reporter);
  my $reporter_args = @_ ? shift : 0;
#use Devel::Peek; Dump($reporter_args);

  X86::Disasm::init($options, $reporter, $reporter_args);
  $self;
}

sub DESTROY {
  X86::Disasm::x86_cleanup();
}

# a utility function used to set a list from a scalar and a hash where
# the scalar is an index to the set keys - esentially we unpack a boolean
sub set_generic {
  my $value = shift;
  my $hash = shift;

  my $list = [];
  return $list unless (defined $value and defined $hash);

  foreach my $key (sort keys %$hash) {
    if ($key) {
      push @$list, $hash->{$key} if ($value & $key);
    } else {
      push @$list, $hash->{$key} if not $value;
    }
  }
  return $list;
}

sub disassemble {
# returns a list of instructions with sub-lists of operands
  my $self = shift;
  my $buffer = shift;
  my $buf_rva = shift;
  my $offset = shift;
  my $format = shift;

  if (ref $buffer and ref $buffer eq 'SCALAR') {
    $buffer = $$buffer;
  }
  my $buf_len = length $buffer;

  my $data;
  my $insn = X86::Disasm::Insn->new;

  while ($offset < $buf_len) {
    my $retval = X86::Disasm::x86_disasm($buffer, $buf_len, $buf_rva, $offset, $insn);

    if ($retval) {
      my $insn_lol = $insn->lol($format);
      push @$data, $insn_lol;
      $offset += $retval;
    } else {
      push @$data, $offset;
      $offset++;
    }
  }
  return $data;
}

sub disassemble_list {
# returns a list of instructions and operands
  my $self = shift;
  my $buffer = shift;
  my $buf_rva = shift;
  my $offset = shift;
  my $format = shift;

  if (ref $buffer and ref $buffer eq 'SCALAR') {
    $buffer = $$buffer;
  }
  my $buf_len = length $buffer;

  my $data;
  my $insn = X86::Disasm::Insn->new;

  while ($offset < $buf_len) {
    my $retval = X86::Disasm::x86_disasm($buffer, $buf_len, $buf_rva, $offset, $insn);

    if ($retval) {
      my $insn_list = $insn->list($format);
      push @$data, $insn_list;
      $offset += $retval;
    } else {
      push @$data, $offset;
      $offset++;
    }
  }
  return $data;
}

sub disassemble_hash {
# returns a hash representing the full disassembly
  my $self = shift;
  my $buffer = shift;
  my $buf_rva = shift;
  my $offset = shift;

  if (ref $buffer and ref $buffer eq 'SCALAR') {
    $buffer = $$buffer;
  }
  my $buf_len = length $buffer;

  my $data;
  my $insn = X86::Disasm::Insn->new;

  while ($offset < $buf_len) {
    my $retval = X86::Disasm::x86_disasm($buffer, $buf_len, $buf_rva, $offset, $insn);

    if ($retval) {
      my $insn_hash = $insn->hash;
      push @$data, $insn_hash;
      $offset += $retval;
    } else {
      push @$data, $offset;
      $offset++;
    }
  }
  return $data;
}

sub disassemble_range {
  my $self = shift;
  my $buffer = shift;
  my $buf_rva = shift;
  my $offset = shift;
  my $len = shift;
  my $callback = shift;
  my $callback_data = shift;

  if (ref $buffer and ref $buffer eq 'SCALAR') {
    $buffer = $$buffer;
  }
  my $retval = X86::Disasm::disasm_range($buffer, $buf_rva, $offset, $len, $callback, $callback_data);
  return $retval;
}

sub disassemble_forward {
  my $self = shift;
  my $buffer = shift;
  my $buf_rva = shift;
  my $offset = shift;
  my $callback = shift;
  my $callback_data = shift;
  my $resolver = shift;
  my $resolver_data = shift;

  if (ref $buffer and ref $buffer eq 'SCALAR') {
    $buffer = $$buffer;
  }
  my $buf_len = length($buffer);

  my $retval = X86::Disasm::disasm_forward($buffer, $buf_len, $buf_rva, $offset, $callback, $callback_data, $resolver, $resolver_data);
  return $retval;
}

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use X86::Disasm ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.

our %EXPORT_TAGS = ( 'all' => [ qw(
$x86_asm_format
$x86_asm_format_enum
$x86_options
$x86_op_foreach_type
$x86_report_codes
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.02';

require XSLoader;
XSLoader::load('X86::Disasm', $VERSION);

# Preloaded methods go here.

1;

__END__

=pod

=head1 NAME

X86::Disasm - Perl extension to wrap libdisasm - an X86 Disassembler

=head1 SYNOPSIS

  use X86::Disasm ':all';

  my $buffer = "\x8d\x4c\x24\x04\x83\xe4\xf0\xff\x71\xfc\x55\x89\xe5\x51";
  my $buf_rva = 0;
  my $offset = 0;

  my $disasm = X86::Disasm->new;

  my $data = $disasm->disassemble($buffer, $buf_rva, $offset, $x86_asm_format_enum->{$syntax});

=head1 DESCRIPTION

X86::Disasm provides a Perl interface to the C X86 disassembler library, libdisasm. See http://bastard.sourceforge.net/libdisasm.html

=head1 EXPORT

None by default.

  our %EXPORT_TAGS = ( 'all' => [ qw(
  $x86_asm_format
  $x86_asm_format_enum
  $x86_options
  $x86_op_foreach_type
  $x86_report_codes
  ) ] );

=head1 METHODS

=head2 new

  my $disasm = X86::Disasm->new($options, $reporter, $reporter_args);

All arguments are optional. 

C<$options> is defined by the hash

  our $x86_options = {	# these can be ORed together 
	0 => "opt_none",
	1 => "opt_ignore_nulls",  # ignore sequences of > 4 NULL bytes
	2 => "opt_16_bit",	  # 16-bit/DOS disassembly
	4 => "opt_att_mnemonics", # use AT&T syntax names for alternate opcode mnemonics
  };

If supplied, C<$reporter> must be a code reference.

If supplied, C<$reporter_args> must be a hash reference.

=head2 disassemble

  my $data = $disasm->disassemble($buffer, $buf_rva, $offset, $x86_asm_format_enum->{$syntax});

This method presents the instructions as a list of lists. Each instruction
is the first element of the sub-list; subsequent elements are the associated
operands.

=head2 disassemble_list

  my $data = $disasm->disassemble_list($buffer, $buf_rva, $offset, $x86_asm_format_enum->{$syntax});

This method presents the instructions as a list. Each instruction is presented
as a string.

=head2 disassemble_hash

  my $data = $disasm->disassemble_hash($buffer, $buf_rva, $offset);

This method presents the instructions as a a list of hashes. Each instruction
is totally deconstructed in to the hash - and provides a full representation
of the information.

=head2 disassemble_range

  $disasm->disassemble_range($buffer, $buf_rva, $offset, $length, $callback_ref, $callback_data);

This method disassembles the range of instructions from $offset for $length 
bytes. The supplied calback can be used to do *something* with the 
instructions.

=head2 disassemble_forward

  my $retval = $disasm->disassemble_forward($buffer, $buf_rva, $offset, $callback_ref, $callback_data, $resolver_ref, $resolver_data);

The disassembly in this case starts at 'offset', and proceeds forward following
the flow of execution for the disassembled code. This means that when a jump,
call, or conditional jump is encountered, disassemble_forward recurses, using
the offset of the target of the jump or call as the 'offset' argument. When
a jump or return is encountered, disassemble_forward returns, allowing its
caller [either the application, or an outer invocation of disassemble_forward]
to continue.

=head1 SEE ALSO

If you use Debian and install libdisasm0 and libdisasm-dev
then the following are a useful supplement to this documentation.

/usr/include/libdis.h

/usr/share/doc/libdisasm-dev/libdisasm.txt.gz

The latest version of this Perl module is available from https://sourceforge.net/projects/x86disasm/

=head1 AUTHOR

Bob Wilkinson, E<lt>bob@fourtheye.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Bob Wilkinson

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/X86/Disasm/Op.pm
package X86::Disasm::Op;

use 5.008000;
use strict;
use warnings;

use X86::Disasm::Reg ':all';

require Exporter;

our @ISA = qw(Exporter);

our $x86_op_type = {	# mutually exclusive 
	0 => "op_unused",	 # empty/unused operand: should never occur
	1 => "op_register",	 # CPU register
	2 => "op_immediate",	 # Immediate Value
	3 => "op_relative_near",	 # Relative offset from IP
	4 => "op_relative_far",	 # Relative offset from IP
	5 => "op_absolute",	 # Absolute address (ptr16:32)
	6 => "op_expression",	 # Address expression (scale/index/base/disp)
	7 => "op_offset",	 # Offset from start of segment (m32)
};

our $x86_op_datatype = {	# these use Intel's lame terminology 
	1 => "op_byte",	 # 1 byte integer
	2 => "op_word",	 # 2 byte integer
	3 => "op_dword",	 # 4 byte integer
	4 => "op_qword",	 # 8 byte integer
	5 => "op_dqword",	 # 16 byte integer
	6 => "op_sreal",	 # 4 byte real (single real)
	7 => "op_dreal",	 # 8 byte real (double real)
	8 => "op_extreal",	 # 10 byte real (extended real)
	9 => "op_bcd",	 # 10 byte binary-coded decimal
	10 => "op_ssimd",	 # 16 byte : 4 packed single FP (SIMD, MMX)
	11 => "op_dsimd",	 # 16 byte : 2 packed double FP (SIMD, MMX)
	12 => "op_sssimd",	 # 4 byte : scalar single FP (SIMD, MMX)
	13 => "op_sdsimd",	 # 8 byte : scalar double FP (SIMD, MMX)
	14 => "op_descr32",	 # 6 byte Intel descriptor 2:4
	15 => "op_descr16",	 # 4 byte Intel descriptor 2:2
	16 => "op_pdescr32",	 # 6 byte Intel pseudo-descriptor 32:16
	17 => "op_pdescr16",	 # 6 byte Intel pseudo-descriptor 8:24:16
	18 => "op_bounds16",	 # signed 16:16 lower:upper bounds
	19 => "op_bounds32",	 # signed 32:32 lower:upper bounds
	20 => "op_fpuenv16",	 # 14 byte FPU control/environment data
	21 => "op_fpuenv32",	 # 28 byte FPU control/environment data
	22 => "op_fpustate16",	 # 94 byte FPU state (env & reg stack)
	23 => "op_fpustate32",	 # 108 byte FPU state (env & reg stack)
	24 => "op_fpregset",	 # 512 bytes: register set
	25 => "op_fpreg",	 # FPU register
	0xFF => "op_none",	 # operand without a datatype (INVLPG)
};

our $x86_op_access = {	# ORed together 
	1 => "op_read",
	2 => "op_write",
	4 => "op_execute",
};

our $x86_op_flags = {	# ORed together, but segs are mutually exclusive 
	1 => "op_signed",	 # signed integer
	2 => "op_string",	 # possible string or array
	4 => "op_constant",	 # symbolic constant
	8 => "op_pointer",	 # operand points to a memory address
	0x010 => "op_sysref",	 # operand is a syscall number
	0x020 => "op_implied",	 # operand is implicit in the insn
	0x40 => "op_hardcode",	 # operand is hardcoded in insn definition
	0x100 => "op_es_seg",	 # ES segment override
	0x200 => "op_cs_seg",	 # CS segment override
	0x300 => "op_ss_seg",	 # SS segment override
	0x400 => "op_ds_seg",	 # DS segment override
	0x500 => "op_fs_seg",	 # FS segment override
	0x600 => "op_gs_seg",	 # GS segment override
};

sub _union {
  my $operand = shift;

  my $datatype = $operand->datatype;
  my $flags = $operand->flags;

  my $union;
  if (defined $x86_op_datatype->{$datatype}) {
    if (defined $x86_op_flags->{$flags}) {
      if ($x86_op_flags->{$flags} && 1) { # signed
        if ($x86_op_datatype->{$datatype} eq 'op_byte') {
          $union = $operand->sbyte;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_word') {
          $union = $operand->sword;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_dword') {
          $union = $operand->sdword;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_qword') {
          $union = $operand->sqword;
        }
      } else {
        if ($x86_op_datatype->{$datatype} eq 'op_byte') {
          $union = $operand->byte;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_word') {
          $union = $operand->word;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_dword') {
          $union = $operand->dword;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_qword') {
          $union = $operand->qword;
        } elsif ($x86_op_datatype->{$datatype} eq 'op_dqword') {
          $union = $operand->dqword;
        }
      }
    }
    if ($x86_op_datatype->{$datatype} eq 'op_sreal') {
      $union = $operand->sreal;
    } elsif ($x86_op_datatype->{$datatype} eq 'op_double') {
      $union = $operand->dreal;
    } elsif ($x86_op_datatype->{$datatype} eq 'op_extreal') {
      $union = $operand->extreal;
    } elsif ($x86_op_datatype->{$datatype} eq 'op_bcd') {
      $union = $operand->bcd;
    } elsif ($x86_op_datatype->{$datatype} eq 'op_simd') {
      $union = $operand->simd;
    } elsif ($x86_op_datatype->{$datatype} eq 'op_fpuenv') {
      $union = $operand->fpuenv;
    }
  }
  return $union;
}

sub set_access {
  my $access = shift;

  return X86::Disasm::set_generic($access, $x86_op_access);
}

sub set_flags {
  my $flags = shift;

  return X86::Disasm::set_generic($flags, $x86_op_flags);
}

sub set {
  my $operand = shift;

  my $op_hash;
  my $type = $operand->type;
  $op_hash->{type} = $x86_op_type->{$operand->type};
  $op_hash->{datatype} = $x86_op_datatype->{$operand->datatype};
  $op_hash->{access} = set_access($operand->access);
  $op_hash->{flags} = set_flags($operand->flags);
  $op_hash->{union} = X86::Disasm::Op::_union($operand);
  $op_hash->{offset} = sprintf("0x%x", $operand->offset);
  if ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_register')) {
    $op_hash->{reg} = X86::Disasm::Reg::set($operand->reg);
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_expression')) {
    my $expression = $operand->expression;
    $op_hash->{expression} = {};
    $op_hash->{expression}->{scale} = $expression->scale;
    $op_hash->{expression}->{index} = X86::Disasm::Reg::set($expression->index);
    $op_hash->{expression}->{base} = X86::Disasm::Reg::set($expression->base);
    $op_hash->{expression}->{disp} = sprintf("0x%x", $expression->disp);
    $op_hash->{expression}->{disp_sign} = ord($expression->disp_sign);
    $op_hash->{expression}->{disp_size} = ord($expression->disp_size);
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_immediate')) {
warn "IMMEDIATE";
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_relative_near')) {
    $op_hash->{relative_near} = ord($operand->relative_near);
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_relative_far')) {
    $op_hash->{relative_farg} = ord($operand->relative_near);
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_absolute')) {
    my $absolute = $operand->absolute;
    $op_hash->{absolute} = {};
    $op_hash->{absolute}->{segment} = $absolute->segment;
    $op_hash->{absolute}->{off16} = $absolute->off16;
    $op_hash->{absolute}->{off32} = $absolute->off32;
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_expression')) {
warn "EXPRESSION";
  } elsif ((defined $x86_op_type->{$type}) and ($x86_op_type->{$type} eq 'op_offset')) {
warn "OFFSET";
  } elsif (defined $x86_op_type->{$type}) {
warn "TYPE is ",$x86_op_type->{$type};
  }
  return $op_hash;
}

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use X86::Disasm ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.

our %EXPORT_TAGS = ( 'all' => [ qw(
$x86_op_type
$x86_op_datatype
$x86_op_access
$x86_op_flags
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

1;

__END__

# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

X86::Disasm::Op - Perl extension to wrap operands from libdisasm

=head1 SYNOPSIS

  use X86::Disasm::Op ':all';

=head1 DESCRIPTION

X86::Disasm::Op provides a Perl interface to the operands in the C X86 disassembler library, libdisasm. See http://bastard.sourceforge.net/libdisasm.html

=head2 EXPORT

None by default.

  our %EXPORT_TAGS = ( 'all' => [ qw(
  $x86_op_type
  $x86_op_datatype
  $x86_op_access
  $x86_op_flags
  ) ] );

=head1 SEE ALSO

  X86::Disasm
  X86::Disasm::Insn
  X86::Disasm::Reg

If you use Debian and install libdisasm0 and libdisasm-dev
then the following are a useful supplement to this documentation.

/usr/include/libdis.h

/usr/share/doc/libdisasm-dev/libdisasm.txt.gz

The latest version of this Perl module is available from https://sourceforge.net/projects/x86disasm/

=head1 AUTHOR

Bob Wilkinson, E<lt>bob@fourtheye.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Bob Wilkinson

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut
